src/panels/Scene/SceneTree/treeUtil.ts

```diff
@@ -0,0 +1,71 @@
+import { TreeNodeData } from '@mantine/core';
+
+import { Properties, PropertyOwners, Uri } from '@/types/types';
+import {
+  isPropertyOwnerHidden,
+  isSceneGraphNodeVisible
+} from '@/util/propertyTreeHelpers';
+
+export const GroupPrefixKey = '/groups/';
+
+export function isGroup(treeNodeDataValue: string): boolean {
+  return treeNodeDataValue.startsWith(GroupPrefixKey);
+}
+
+export function isGroupNode(node: TreeNodeData): boolean {
+  return isGroup(node.value);
+}
+
+export interface SceneTreeFilterProps {
+  showOnlyVisible: boolean;
+  showHiddenNodes: boolean;
+  tags: string[];
+}
+
+export function filterTreeData(
+  nodes: TreeNodeData[],
+  filter: SceneTreeFilterProps,
+  properties: Properties,
+  propertyOwners: PropertyOwners
+): TreeNodeData[] {
+  const filteredNodes = nodes
+    .map((node) => {
+      const newNode = { ...node };
+      if (isGroupNode(newNode)) {
+        // Groups => filter children
+        newNode.children = filterTreeData(
+          newNode.children || [],
+          filter,
+          properties,
+          propertyOwners
+        );
+        if (newNode.children.length === 0) {
+          // Don't show empty groups
+          return null;
+        }
+        return newNode;
+      } else {
+        function shouldShowSceneGraphNode(uri: Uri) {
+          let shouldShow = true;
+          if (filter.showOnlyVisible) {
+            shouldShow &&= isSceneGraphNodeVisible(uri, properties);
+          }
+          if (!filter.showHiddenNodes) {
+            shouldShow &&= !isPropertyOwnerHidden(uri, properties);
+          }
+          if (shouldShow && filter.tags.length > 0) {
+            shouldShow &&= filter.tags.some((tag) =>
+              propertyOwners[newNode.value]?.tags.includes(tag)
+            );
+          }
+          return shouldShow;
+        }
+
+        // PropertyOwners (scene graph nodes), may be filtered out based on settings
+        return shouldShowSceneGraphNode(newNode.value) ? newNode : null;
+      }
+    })
```

Comment:
Consider breaking this out to its own function. Lots of stuff happening with map and filter which makes it a bit hard to read